src/main/scala/tech/sourced/berserker/SparkDriver.scala

```diff
@@ -66,65 +56,62 @@ object SparkDriver {
+        new SivaUnpacker(siva.getAbsolutePath).unpack(unpackDir)
+        (sivaFile.substring(0, sivaFile.lastIndexOf('.')), unpackDir)
+      } //RDD["sivaFileName sivaUnpackDir"]
 
     // iterate every un-packed .siva
     val intermediatePerFile = unpacked
       .mapPartitions(partition => {
         val log = Logger.getLogger(s"Stage: single partition")
-
-        val enryService = EnryService(enryHost, enryPort, grpcMaxMsgSize)
         val bblfshService = BblfshClient(bblfshHost, bblfshPort, grpcMaxMsgSize)
-        log.info(s"Connecting to Enry server: $enryHost:$enryPort")
         log.info(s"Connecting to Bblfsh server: $bblfshHost:$bblfshPort")
 
         partition
-          .flatMap { sivaFileNameAndUnpackedDir =>
-            val log = Logger.getLogger(s"Stage: process single repo '$sivaFileNameAndUnpackedDir'")
-            val (sivaFileName, sivaUnpackDir) = split(sivaFileNameAndUnpackedDir)
-
+          .flatMap { case (sivaFileName, sivaUnpackDir) =>
+            val log = Logger.getLogger(s"Stage: process single")
             log.info(s"Processing repository in $sivaUnpackDir")
-
             JGitFileIterator(sivaUnpackDir, sivaFileName, confBroadcast.value.value)
           }
-          .filter { case (_, treeWalk, _) =>
+          .filter { case (_, treeWalk, _, _) =>
             treeWalk.getFileMode(0) == FileMode.REGULAR_FILE || treeWalk.getFileMode(0) == FileMode.EXECUTABLE_FILE
             //TODO(bzz): skip big well-known binaries .apk and .jar
           }
-          .map { case (initHash, treeWalk, ref) =>
+          .map { case (initHash, tree, ref, config) =>
             val log = Logger.getLogger(s"Stage: detecting a language")
-            val path = treeWalk.getPathString
+            val path = tree.getPathString
 
             var content = Array.emptyByteArray
-            var guessed: EnryResponse = null
-            try { // detect language using enry server
-              guessed = enryService.getLanguage(path)
-              if (guessed.status == Status.NEED_CONTENT) {
-                content = RootedRepo.readFile(treeWalk.getObjectId(0), treeWalk.getObjectReader)
-                guessed = enryService.getLanguage(path, content)
+            var guessed = Enry.getLanguageByFilename(path)
+            if (!guessed.safe) {
+              content = RootedRepo.readFile(tree.getObjectId(0), tree.getObjectReader)
+              log.info(s"Detecting lang for: $path using content size:${content.length}")
+              guessed = if (content.isEmpty) {
+                new Guess("", false) //Enry.unknownLanguage
+              } else {
+                Enry.getLanguageByContent(path, content)
               }
-            }  catch {
-              case e: Throwable => log.info(s"Could not detect language for $path, $e")
             }
-            (initHash, treeWalk, ref, path, content, guessed)
+            (initHash, tree, ref, config, path, content, guessed)
```

Comment:
Good point. We shall have a case class for this in spark-api.

For now, if you are do not have a strong opinion, I would prefer keeping it as a tuple, simply because API is not settled yet and it's is easier to modify.

But that's a good point, as soon as structure of API is settled and we understand all the dependencies - we will add a case class.